presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DesugarArrayConstructorSubquery.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.metadata.Metadata;
+import io.prestosql.sql.planner.FunctionCallBuilder;
+import io.prestosql.sql.tree.ArrayConstructor;
+import io.prestosql.sql.tree.CurrentUser;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.ExpressionTreeRewriter;
+import io.prestosql.sql.tree.FunctionCall;
+import io.prestosql.sql.tree.QualifiedName;
+import io.prestosql.sql.tree.SubqueryExpression;
+
+public class DesugarArrayConstructorSubquery
+        extends ExpressionRewriteRuleSet
+{
+    public DesugarArrayConstructorSubquery(Metadata metadata)
+    {
+        super(createRewrite(metadata));
+    }
+
+    private static ExpressionRewriter createRewrite(Metadata metadata)
+    {
+        return (expression, context) -> ExpressionTreeRewriter.rewriteWith(new io.prestosql.sql.tree.ExpressionRewriter<Void>()
+        {
+            @Override
+            public Expression rewriteArrayConstructor(ArrayConstructor node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                if (node.isSubquery()) {
+                    SubqueryExpression subquery = node.getSubquery();
+                    // TODO: wrap single projected node with ARRAY_AGG
```

Comment:
Is this the right place to rewrite the subquery to wrap the projected node with an ARRAY_AGG?

Reply:
Take a look at `TransformUncorrelatedInPredicateSubqueryToSemiJoin`